File: ios-the-aura-den/TheAuraDen/Components/BrandComponents.swift
import SwiftUI

/// The Parisienne logotype, used on brand moments only.
struct LogoWordmark: View {
    var size: CGFloat = 46
    var color: Color = AuraPalette.white

    var body: some View {
        Text(AuraCopy.brandName)
            .font(AuraFont.logo(size))
            .foregroundStyle(color)
            .accessibilityAddTraits(.isHeader)
    }
}

/// The half-ellipse brand arc that crowns the logotype.
nonisolated struct AuraArcShape: Shape {
    var lineWidth: CGFloat = 2

    func path(in rect: CGRect) -> Path {
        let inset = lineWidth / 2
        let width = max(rect.width - lineWidth, 1)
        let height = max(rect.height - inset, 1)

        var unit = Path()
        unit.addArc(
            center: .zero,
            radius: 0.5,
            startAngle: .degrees(180),
            endAngle: .degrees(360),
            clockwise: false
        )

        let transform = CGAffineTransform(translationX: rect.midX, y: rect.maxY)
            .scaledBy(x: width, y: height * 2)
        return unit.applying(transform)
    }
}

/// Stroked brand arc, sized by its frame.
struct AuraArc: View {
    var color: Color = AuraPalette.yellow
    var lineWidth: CGFloat = 2

    var body: some View {
        AuraArcShape(lineWidth: lineWidth)
            .stroke(color, style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
    }
}

enum LogoSize {
    case large
    case medium
    case compact

    var script: CGFloat {
        switch self {
        case .large: 58
        case .medium: 38
        case .compact: 26
        }
    }

    var arcWidth: CGFloat {
        switch self {
        case .large: 86
        case .medium: 58
        case .compact: 40
        }
    }

    var theSize: CGFloat {
        switch self {
        case .large: 16
        case .medium: 12
        case .compact: 9
        }
    }

    var denSize: CGFloat {
        switch self {
        case .large: 14
        case .medium: 11
        case .compact: 8
        }
    }

    var arcSpacing: CGFloat {
        self == .large ? 10 : 4
    }

    var arcStroke: CGFloat {
        self == .large ? 2.5 : 1.5
    }
}

/// "The Aura Den" logotype: brand arc over the script wordmark, exactly as on Android.
struct AuraLogo: View {
    var size: LogoSize = .large
    var textColor: Color = AuraPalette.white
    var scriptColor: Color = AuraPalette.yellow
    var arcColor: Color = AuraPalette.yellow

    var body: some View {
        VStack(spacing: 0) {
            AuraArc(color: arcColor, lineWidth: size.arcStroke)
                .frame(width: size.arcWidth, height: size.arcWidth / 2.6)
                .padding(.bottom, size.arcSpacing)

            Text("The")
                .font(AuraFont.displaySmall(size.theSize))
                .foregroundStyle(textColor)

            Text("Aura")
                .font(AuraFont.logo(size.script))
                .foregroundStyle(scriptColor)
                .padding(.vertical, -size.script * 0.16)

            Text("D E N")
                .font(AuraFont.labelMedium(size.denSize))
                .tracking(size.denSize * 0.35)
                .foregroundStyle(textColor)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(AuraCopy.brandName)
        .accessibilityAddTraits(.isHeader)
    }
}

/// Very subtle dotted texture inspired by the brand board.
nonisolated struct BrandTexture: View {
    var color: Color = AuraPalette.navy
    var opacity: Double = 0.35

    var body: some View {
        Canvas { context, size in
            let step = size.width / 7
            let dot: CGFloat = 0.8
            var y = step * 0.6
            while y < size.height {
                var x = step * 0.5
                while x < size.width {
                    let rect = CGRect(
                        x: x - dot,
                        y: y - dot,
                        width: dot * 2,
                        height: dot * 2
                    )
                    context.fill(Path(ellipseIn: rect), with: .color(color.opacity(opacity)))
                    x += step
                }
                y += step
            }
        }
        .allowsHitTesting(false)
    }
}

/// The signature arch shape from the brand board — a rounded top on a straight base.
nonisolated struct ArchShape: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        let radius = rect.width / 2
        path.move(to: CGPoint(x: rect.minX, y: rect.maxY))
        path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + radius))
        path.addArc(
            center: CGPoint(x: rect.midX, y: rect.minY + radius),
            radius: radius,
            startAngle: .degrees(180),
            endAngle: .degrees(0),
            clockwise: false
        )
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
        path.closeSubpath()
        return path
    }
}

/// Brand header used at the top of every screen: deep blue block with soft rounded bottom,
/// optional back arrow and optional trailing action.
struct AuraHeader<Trailing: View, Content: View>: View {
    let title: String
    var eyebrow: String?
    var subtitle: String?
    var onBack: (() -> Void)?
    var containerColor: Color = AuraPalette.blue
    var titleColor: Color = AuraPalette.white
    @ViewBuilder var trailing: Trailing
    @ViewBuilder var content: Content

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(alignment: .center, spacing: 0) {
                if let onBack {
                    Button {
                        AuraHaptics.tap()
                        onBack()
                    } label: {
                        Image(systemName: "arrow.left")
                            .font(.system(size: 20, weight: .medium))
                            .foregroundStyle(titleColor)
                            .frame(width: 48, height: 48)
                            .contentShape(.rect)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Regresar")
                } else {
                    Spacer().frame(width: 8)
                }

                VStack(alignment: .leading, spacing: 0) {
                    if let eyebrow {
                        Text(eyebrow.uppercased())
                            .font(AuraFont.labelSmall())
                            .tracking(1.4)
                            .foregroundStyle(AuraPalette.yellow)
                            .padding(.bottom, 4)
                    }
                    Text(title)
                        .font(AuraFont.headlineMedium())
                        .foregroundStyle(titleColor)
                        .fixedSize(horizontal: false, vertical: true)
                    if let subtitle {
                        Text(subtitle)
                            .font(AuraFont.bodyMedium())
                            .foregroundStyle(titleColor.opacity(0.82))
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.top, 3)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)

                trailing
                    .padding(.leading, 12)
            }

            content
                .padding(.horizontal, 8)
                .padding(.top, 18)
        }
        .padding(.leading, 8)
        .padding(.trailing, 16)
        .padding(.top, 8)
        .padding(.bottom, 22)
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.top, safeAreaTop)
        .background(
            containerColor,
            in: .rect(bottomLeadingRadius: 28, bottomTrailingRadius: 28)
        )
    }

    private var safeAreaTop: CGFloat {
        #if canImport(UIKit)
        let scenes = UIApplication.shared.connectedScenes
        let window = scenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first { $0.isKeyWindow }
        return window?.safeAreaInsets.top ?? 47
        #else
        return 47
        #endif
    }
}

extension AuraHeader where Trailing == EmptyView, Content == EmptyView {
    init(
        title: String,
        eyebrow: String? = nil,
        subtitle: String? = nil,
        onBack: (() -> Void)? = nil,
        containerColor: Color = AuraPalette.blue,
        titleColor: Color = AuraPalette.white
    ) {
        self.init(
            title: title,
            eyebrow: eyebrow,
            subtitle: subtitle,
            onBack: onBack,
            containerColor: containerColor,
            titleColor: titleColor,
            trailing: { EmptyView() },
            content: { EmptyView() }
        )
    }
}

extension AuraHeader where Trailing == EmptyView {
    init(
        title: String,
        eyebrow: String? = nil,
        subtitle: String? = nil,
        onBack: (() -> Void)? = nil,
        containerColor: Color = AuraPalette.blue,
        titleColor: Color = AuraPalette.white,
        @ViewBuilder content: () -> Content
    ) {
        self.init(
            title: title,
            eyebrow: eyebrow,
            subtitle: subtitle,
            onBack: onBack,
            containerColor: containerColor,
            titleColor: titleColor,
            trailing: { EmptyView() },
            content: content
        )
    }
}

extension AuraHeader where Content == EmptyView {
    init(
        title: String,
        eyebrow: String? = nil,
        subtitle: String? = nil,
        onBack: (() -> Void)? = nil,
        containerColor: Color = AuraPalette.blue,
        titleColor: Color = AuraPalette.white,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.init(
            title: title,
            eyebrow: eyebrow,
            subtitle: subtitle,
            onBack: onBack,
            containerColor: containerColor,
            titleColor: titleColor,
            trailing: trailing,
            content: { EmptyView() }
        )
    }
}

/// Small decorative arch + wordmark for card corners, matching the membership cards.
struct AuraCardMark: View {
    var tint: Color = AuraPalette.blue

    var body: some View {
        VStack(spacing: 0) {
            AuraArc(color: AuraPalette.yellow, lineWidth: 1.5)
                .frame(width: 34, height: 13)
                .padding(.bottom, 3)
            Text("The")
                .font(AuraFont.displaySmall(9))
                .foregroundStyle(tint)
            Text("Aura")
                .font(AuraFont.logo(22))
                .foregroundStyle(tint)
                .padding(.vertical, -4)
            Text("D E N")
                .font(AuraFont.labelMedium(7))
                .tracking(2.4)
                .foregroundStyle(tint)
        }
    }
}

/// Deep navy header with the arch motif behind the content.
struct AuraBrandHeader<Content: View>: View {
    var showsLogo: Bool = true
    @ViewBuilder var content: Content

    var body: some View {
        ZStack(alignment: .top) {
            LinearGradient(
                colors: [AuraPalette.blue, AuraPalette.navyDeep],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
            .overlay(alignment: .topTrailing) {
                ArchShape()
                    .fill(AuraPalette.white.opacity(0.06))
                    .frame(width: 190, height: 240)
                    .offset(x: 60, y: -70)
                    .allowsHitTesting(false)
            }
            .overlay(alignment: .bottomLeading) {
                Circle()
                    .fill(AuraPalette.yellow.opacity(0.14))
                    .frame(width: 150, height: 150)
                    .offset(x: -60, y: 70)
                    .allowsHitTesting(false)
            }

            VStack(alignment: .leading, spacing: 14) {
                if showsLogo {
                    LogoWordmark(size: 34)
                }
                content
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 20)
            .padding(.top, 12)
            .padding(.bottom, 26)
        }
        .clipShape(.rect(bottomLeadingRadius: 30, bottomTrailingRadius: 30))
        .ignoresSafeArea(edges: .top)
    }
}

/// Small "demo" marker so nobody mistakes the prototype for the shipped product.
struct DemoBadge: View {
    var body: some View {
        Text("Demo")
            .font(AuraFont.labelSmall())
            .tracking(1.2)
            .foregroundStyle(AuraPalette.navy)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(AuraPalette.yellow, in: .capsule)
    }
}

/// Circular remote avatar with a warm placeholder.
struct AuraAvatar: View {
    let imageURL: String
    var size: CGFloat = 52
    var ringColor: Color = AuraPalette.white

    var body: some View {
        Circle()
            .fill(AuraPalette.sandSoft)
            .frame(width: size, height: size)
            .overlay {
                AsyncImage(url: URL(string: imageURL)) { phase in
                    switch phase {
                    case .success(let image):
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fill)
                    default:
                        Image(systemName: "person.fill")
                            .font(.system(size: size * 0.42))
                            .foregroundStyle(AuraPalette.sand)
                    }
                }
                .allowsHitTesting(false)
            }
            .clipShape(.circle)
            .overlay {
                Circle().strokeBorder(ringColor.opacity(0.7), lineWidth: 2)
            }
    }
}

File: android/app/src/main/java/com/rork/theauraden/ui/components/BrandComponents.kt
package com.rork.theauraden.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.theauraden.ui.theme.AuraBlue
import com.rork.theauraden.ui.theme.AuraWhite
import com.rork.theauraden.ui.theme.AuraYellow
import com.rork.theauraden.ui.theme.BodyFontFamily
import com.rork.theauraden.ui.theme.DisplayFontFamily
import com.rork.theauraden.ui.theme.EyebrowStyle
import com.rork.theauraden.ui.theme.LogoFontFamily

/** The simple brand arch that sits above the logotype. */
@Composable
fun AuraArc(
    modifier: Modifier = Modifier,
    color: Color = AuraYellow,
    strokeWidth: Dp = 2.dp
) {
    Canvas(modifier = modifier) {
        val stroke = strokeWidth.toPx()
        drawArc(
            color = color,
            startAngle = 180f,
            sweepAngle = 180f,
            useCenter = false,
            topLeft = Offset(stroke / 2f, stroke / 2f),
            size = Size(size.width - stroke, size.height * 2f - stroke),
            style = Stroke(width = stroke, cap = StrokeCap.Round)
        )
    }
}

enum class LogoSize { Large, Medium, Compact }

/** "The Aura Den" logotype: script wordmark under the arch. */
@Composable
fun AuraLogo(
    modifier: Modifier = Modifier,
    size: LogoSize = LogoSize.Large,
    textColor: Color = AuraWhite,
    scriptColor: Color = AuraYellow,
    arcColor: Color = AuraYellow
) {
    val scriptSize = when (size) {
        LogoSize.Large -> 58.sp
        LogoSize.Medium -> 38.sp
        LogoSize.Compact -> 26.sp
    }
    val arcWidth = when (size) {
        LogoSize.Large -> 86.dp
        LogoSize.Medium -> 58.dp
        LogoSize.Compact -> 40.dp
    }
    val theSize = when (size) {
        LogoSize.Large -> 16.sp
        LogoSize.Medium -> 12.sp
        LogoSize.Compact -> 9.sp
    }
    val denSize = when (size) {
        LogoSize.Large -> 14.sp
        LogoSize.Medium -> 11.sp
        LogoSize.Compact -> 8.sp
    }

    Column(modifier = modifier, horizontalAlignment = Alignment.CenterHorizontally) {
        AuraArc(
            modifier = Modifier
                .width(arcWidth)
                .height(arcWidth / 2.6f),
            color = arcColor,
            strokeWidth = if (size == LogoSize.Large) 2.5.dp else 1.5.dp
        )
        Spacer(Modifier.height(if (size == LogoSize.Large) 10.dp else 4.dp))
        Text(
            text = "The",
            fontFamily = DisplayFontFamily,
            fontSize = theSize,
            color = textColor,
            textAlign = TextAlign.Center
        )
        Text(
            text = "Aura",
            fontFamily = LogoFontFamily,
            fontSize = scriptSize,
            color = scriptColor,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = if (size == LogoSize.Large) 0.dp else 0.dp)
        )
        Text(
            text = "D E N",
            fontFamily = BodyFontFamily,
            fontWeight = FontWeight.Medium,
            fontSize = denSize,
            letterSpacing = denSize * 0.35f,
            color = textColor,
            textAlign = TextAlign.Center
        )
    }
}

/**
 * Brand header used at the top of every screen: deep blue block with soft rounded bottom,
 * optional back arrow and optional trailing action.
 */
@Composable
fun AuraHeader(
    title: String,
    modifier: Modifier = Modifier,
    eyebrow: String? = null,
    subtitle: String? = null,
    onBack: (() -> Unit)? = null,
    containerColor: Color = AuraBlue,
    titleColor: Color = AuraWhite,
    trailing: @Composable (() -> Unit)? = null,
    content: @Composable (() -> Unit)? = null
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .background(
                color = containerColor,
                shape = RoundedCornerShape(bottomStart = 28.dp, bottomEnd = 28.dp)
            )
    ) {
        Column(
            modifier = Modifier
                .statusBarsPadding()
                .padding(start = 8.dp, end = 16.dp, top = 8.dp, bottom = 22.dp)
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                if (onBack != null) {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.Rounded.ArrowBack,
                            contentDescription = "Regresar",
                            tint = titleColor
                        )
                    }
                } else {
                    Spacer(Modifier.width(8.dp))
                }
                Column(modifier = Modifier.weight(1f)) {
                    if (eyebrow != null) {
                        Text(
                            text = eyebrow.uppercase(),
                            style = EyebrowStyle,
                            color = AuraYellow
                        )
                        Spacer(Modifier.height(4.dp))
                    }
                    Text(
                        text = title,
                        style = MaterialTheme.typography.headlineMedium,
                        color = titleColor
                    )
                    if (subtitle != null) {
                        Spacer(Modifier.height(3.dp))
                        Text(
                            text = subtitle,
                            style = MaterialTheme.typography.bodyMedium,
                            color = titleColor.copy(alpha = 0.82f)
                        )
                    }
                }
                if (trailing != null) {
                    Spacer(Modifier.width(12.dp))
                    trailing()
                }
            }
            if (content != null) {
                Spacer(Modifier.height(18.dp))
                Box(modifier = Modifier.padding(horizontal = 8.dp)) { content() }
            }
        }
    }
}

/** Small decorative arch + wordmark for card corners, matching the membership cards. */
@Composable
fun AuraCardMark(modifier: Modifier = Modifier, tint: Color = AuraBlue) {
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        AuraArc(
            modifier = Modifier
                .width(34.dp)
                .height(13.dp),
            color = AuraYellow,
            strokeWidth = 1.5.dp
        )
        Spacer(Modifier.height(3.dp))
        Text(text = "The", fontFamily = DisplayFontFamily, fontSize = 9.sp, color = tint)
        Text(text = "Aura", fontFamily = LogoFontFamily, fontSize = 22.sp, color = tint)
        Text(
            text = "D E N",
            fontFamily = BodyFontFamily,
            fontSize = 7.sp,
            letterSpacing = 2.4.sp,
            color = tint
        )
    }
}


